### PR TITLE
ncm-network: nmstate - add additional route rule parameters

### PR DESCRIPTION
**Pull request by aka7**
provide additional route rule parameters for nmstate config as defined in https://nmstate.io/devel/yaml_api.html#routes

* Why the change is necessary.
allow creation of route rules with all options.
* What backwards incompatibility it may introduce.
none.

### CODE REVIEW

**Comment by aka7** [comment]
@jrha  created this for discussion on if there will a way to create sperate schemas for nmstate only options. something to think about. as discussed before.

this isnt urgent but at some point we may need to create unreachable routes and only way to do this with nmstate is through route-rules.

**Comment by jrha** [comment]
I'll try and do some testing to establish a way of making the schema modular, but otherwise having these options as a separate `nmstate` sub-dict would seem cleanest.

File: ncm-network/src/main/pan/components/network/core-schema.pan [hunk 1]
```diff
@@ -77,6 +77,16 @@ type structure_rule = {
     "table" ? network_valid_routing_table
     @{priority, The priority of the rule over the others. Required by Network Manager when setting routing rules.}
     "priority" ? long(0..0xffffffff)
+    @{nmstate-action used by nmstate module}
```
Comment (stdweird) [inline ncm-network/src/main/pan/components/network/core-schema.pan:80]: i am not in favour of the naming.
at least make a `structure_rule_nmstate` and move them there. 

are these simply ignored by the old network.pm code? if so, i see not much harm.

wrt separate scheam's i can make a PR. i need to work in the core schema anyway to add the IB support.

Comment (aka7) [reply ncm-network/src/main/pan/components/network/core-schema.pan:80]: @stdweird I may need some help/guide from you to move this into the new structure you created, please.

**Comment by jrha** [comment]
@aka7 I've had a look at how to deal with the schema and it's probably easiest for me to open a PR against your branch.

**Comment by stdweird** [comment]
@jrha  @aka7 let me know about the PR. but should be something like

in the profile: 
- set QUATTOR_TYPES_NETWORK_BACKEND to nmstate
- set QUATTOR_TYPES_NETWORK_LEGACY to false 

you should be able to do this for at least all el9 hosts. 
best to compile and diff the profiles with and without these set. it should be identical

once you did this, you can start to customise the backend schemas:
typically you add the same type to both backends, but eg on initiscripts https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/main/pan/components/network/types/network/backend/initscripts.pan you leave it empty
```
type network_very_special {}
```
and in the nmstate backend https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/main/pan/components/network/types/network/backend/nmstate.pan you define what you want
```
type network_very_special {
   woohoo: boolean
```

these special backend templates are included very early in https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/main/pan/components/network/types/network.pan (this is used insetad of the core-legacy schema if you set the previous variables).

then eg in the "shared" route schema https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/main/pan/components/network/types/network/route.pan you add it like
```
type network_route = {
    include network_very_special
```

and now the nmstate profiles will have support for the woohoo attribute

make sure you have a very unique type name for the type you want to include. the should not have the backend in it (even if it is only implemented for a single backend).


(anyway, untested ;)

**Comment by jrha** [comment]
Schema migration in #1720.

**Comment by jrha** [comment]
Rebased and updated to use new schema.

File: ncm-network/src/main/pan/components/network/types/network/backend/nmstate.pan [hunk 2]
```diff
@@ -17,6 +17,19 @@ type structure_network_backend_specific = {
...
+    @{action used by nmstate module}
+    "action" ? choice('blackhole', 'prohibit', 'unreachable')
+    @{state used by nmstate module, Can only set to absent for deleting matching route rules}
+    "state" ? choice('absent')
+    @{iif used by nmstate module, Incoming interface name}
+    "iif" ? string
+    @{fwmark used by nmstate module. Select the fwmark value to match}
+    "fwmark" ? string
+    @{fwmask used by nmstate module. Select the fwmask value to match}
+    "fwmask" ? string
```
[the remark this answers is not in the document]
Comment (jrha) [reply ncm-network/src/main/pan/components/network/types/network/backend/nmstate.pan:30]: I assume `iif` should be one of the interface names in `/system/network/interfaces/` and/or `/hardware/cards/nic/`?

`fwmark` and `fwmask` should apparently be "Integer or hex string".

[the remark this answers is not in the document]
Comment (jrha) [reply ncm-network/src/main/pan/components/network/types/network/backend/nmstate.pan:30]: Okay both `fwmark` and `fwmask` represent unsigned 32-bit integers.

https://github.com/torvalds/linux/blob/ff7afaeca1a15fbeaa2c4795ee806c0667bd77b2/include/net/fib_rules.h#L24

[the remark this answers is not in the document]
Comment (jrha) [reply ncm-network/src/main/pan/components/network/types/network/backend/nmstate.pan:30]: Something like this perhaps?
```pan
string with match(SELF, '^[0-9a-f]{1,8}$')
```

**Comment by aka7** [comment]
thank you so much @jrha , LGTM

**Comment by jrha** [comment]
@aka7 I've locked down the types as per my previous comments, if it all makes sense then I'm happy to try and also get this into 24.10.

**Comment by aka7** [comment]
> @aka7 I've locked down the types as per my previous comments, if it all makes sense then I'm happy to try and also get this into 24.10.

@jrha yes please that make sense.

**Review by jrha — APPROVED** [review]

**Pull request merged**